File: cpp/day6.cpp
#include "tools.h"
#include <iostream>
#include <fstream>
#include <unordered_map>
#include <unordered_set>
#include <cassert>

typedef std::pair<int, int> pi;
typedef std::unordered_map<pi, int, pair_hash> mpi;
typedef std::unordered_set<pi, pair_hash> spi;

struct ParseFile { mpi locs; pi start; };
constexpr int ROWS = 130;
constexpr int COLS = 130;
char mat[ROWS][COLS] {{'.'}};
const pi dirs[4] {{-1, 0}, {0, 1}, {1, 0}, {0, -1}};

auto parseFile() -> ParseFile;
auto solution(pi) -> spi;
auto solution2(const pi&, const mpi&, const spi&) -> int;

auto main() -> int {
    auto [locs, start] = parseFile();
    
    auto pt1 = solution(start);
    auto pt2 = solution2(start, locs, pt1);
    printf("Part 1: %d\nPart 2: %d", (int) pt1.size(), pt2); 
}

auto parseFile() -> ParseFile {
    mpi locs;
    pi start;
    
    std::ifstream file ("input/day6.txt");
    assert(file);
    std::string line;
    
    size_t i {0};
    while (std::getline(file, line)) {    
        for (size_t j {0}; j < ROWS; ++j) {
            if (line[j] == '#' || line[j] == '^') mat[i][j] = line[j];
            auto pair = std::make_pair(i, j);
            if (line[j] == '^') start = pair;
            if (line[j] == '#') locs[pair] = 0;
        }
        
        ++i;         
    }
    
    return {locs, start};
}

auto inbounds(const pi &pos) -> bool {
    return ((pos.first > 0 && pos.first < ROWS - 1) && (pos.second > 0 && pos.second < COLS - 1));
}

auto solution(pi pos) -> spi {
    auto facing = 0;
    spi visited;
    
    while (inbounds(pos)) {
        auto n_pos = dirs[facing] + pos;
        if (mat[n_pos.first][n_pos.second] == '#') {
            facing++;
            facing = facing % 4;
        } else {
            visited.emplace(pos);
            pos = n_pos;
        }
    }
    visited.emplace(pos);
    
    return visited;
}

auto solution2(const pi &start, const mpi &loc, const spi &empty) -> int {
    auto ttl = 0;
    auto locs = loc;

    for (auto &i : empty) {
        auto facing = 0;
        auto pos = start;
        mat[i.first][i.second] = '#';
        locs[i] = 0;

        while (inbounds(pos)) {
            auto n_pos = dirs[facing] + pos;;
            if (arrValue(mat, n_pos) == '#') {
                facing++;
                locs[n_pos]++;
                if (locs[n_pos] == 4) {
                    // std::cout << "Obstacle was at pos " << i << "\n";
                    ttl++;
                    break;
                }
                facing = facing % 4;
            }
            else pos = n_pos;
        }
        mat[i.first][i.second] = '.';
        locs = loc;
    }
    return ttl;
}

File: rust/day6.rs
use std::collections::{HashMap, HashSet};
use std::fs::read_to_string;

type TI = (i32, i32);
type STI = HashSet<TI>;
type MITI = HashMap<i32, TI>;
type MTII = HashMap<TI, i32>;
type Mat = [[char; 130]; 130];

static ROWS: usize = 130;
static COLS: usize = 130;

fn main() {
    let dirs: MITI = [(0, (-1, 0)), (1, (0, 1)), (2, (1, 0)), (3, (0, -1))].into();
    let (pos, mat, locs) = parse_file();
    let visited = solution(&pos, &mat, &dirs);
    let pt2 = solution2(&pos, &mat, &locs, &dirs, &visited);
    println!("Part 1: {}\nPart 2: {}", visited.len(), pt2)
}

fn parse_file() -> (TI, Mat, MTII) {
    let con = read_to_string("../inputs/day6.txt").unwrap();

    let mut start: TI = (0, 0);
    let mut mat: Mat = [['.'; ROWS]; COLS];
    let mut locs: MTII = HashMap::new();

    con.split("\r\n").enumerate().for_each(|(r_ind, row)| {
        row.chars().enumerate().for_each(|(c_ind, col)| {
            mat[r_ind][c_ind] = col;
            if col == '#' {
                locs.insert((r_ind as i32, c_ind as i32), 0);
            } else if col == '^' {
                start = (r_ind as i32, c_ind as i32);
            }
        })
    });

    return (start, mat, locs);
}

fn arr_value(mat: &Mat, loc: &TI) -> char {
    mat[loc.0 as usize][loc.1 as usize]
}

fn add_tups(x: TI, y: TI) -> TI {
    (x.0 + y.0, x.1 + y.1)
}

fn is_within_bounds(pos: TI) -> bool {
    pos.0 > 0 && (pos.0 as usize) < ROWS - 1 && pos.1 > 0 && (pos.1 as usize) < COLS - 1
}

fn solution(pos: &TI, mat: &Mat, dirs: &MITI) -> STI {
    let mut facing = 0;
    let mut pos = *pos;
    let mut visited: STI = HashSet::new();

    while is_within_bounds(pos) {
        let next_pos = add_tups(pos, dirs[&facing]);
        if arr_value(mat, &next_pos) == '#' {
            facing += 1;
            facing = facing % 4
        } else {
            visited.insert(pos);
            pos = next_pos;
        }
    }
    visited.insert(pos);

    visited
}

fn solution2(pos: &TI, mat: &Mat, locs: &MTII, dirs: &MITI, empty: &STI) -> i32 {
    let mut locs = locs.clone();
    let mut mat = *mat;
    let mut ttl = 0;

    for i in empty {
        let mut facing = 0;
        let mut pos = *pos;
        mat[i.0 as usize][i.1 as usize] = '#';
        locs.insert(*i, 0);

        while is_within_bounds(pos) {
            let next_pos = add_tups(pos, dirs[&facing]);
            if arr_value(&mat, &next_pos) == '#' {
                facing += 1;
                if let Some(value) = locs.get_mut(&next_pos) {
                    *value += 1;
                }
                if locs[&next_pos] == 4 {
                    ttl += 1;
                    break;
                }
                facing = facing % 4
            } else {
                pos = next_pos;
            }
        }
        mat[i.0 as usize][i.1 as usize] = '.';
        for value in locs.values_mut() {
            *value = 0;
        }
        locs.remove(&i);
    }

    ttl
}
